### PR TITLE
Port Hacluster tests from Amulet to Zaza

### PR DESCRIPTION
**Pull request by xiywang**
Signed-off-by: Xiyue Wang <celia.wang@canonical.com>

### CODE REVIEW

File: zaza/openstack/charm_tests/hacluster/tests.py [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+                    leader = unit
+        leader_unit = zaza.model.get_unit_from_name(leader)
+        assert leader_unit.workload_status == "active"
+
+        if primary_status["units"][leader].get("subordinates"):
+            for subordinate in primary_status["units"][leader]["subordinates"]:
+                logging.info("Cleaning {}".format(subordinate))
+                __action = "cleanup"
+                action_id = zaza.model.run_action(subordinate, "cleanup")
+                zaza.model.block_until_unit_wl_status(leader, "active")
```
Comment (thedac) [inline zaza/openstack/charm_tests/hacluster/tests.py]: This status check probably does not help us. We are acting on the subordinate in a way that does not affect the primary.

File: zaza/openstack/charm_tests/hacluster/tests.py [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+                logging.info("Cleaning action w/resource {}"
+                        .format(subordinate))
+                params = {'resource': 'res_ks_haproxy'}
+                __action = "cleanup res_ks_haproxy"
+                zaza.model.run_action(subordinate, "cleanup", \
+                        action_params=params)
+                assert "success" in action_id.data["results"]["result"], (
+                    "Set hacluster action {} failed: {}"
+                    .format(__action, action_id.data))
+                zaza.model.block_until_unit_wl_status(leader, "active")
```
Comment (thedac) [inline zaza/openstack/charm_tests/hacluster/tests.py]: Same with this one.

File: zaza/openstack/charm_tests/hacluster/tests.py [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+        # Use the application it is subordinate-to to check workload status
+        if status.get("units") is None and status.get("subordinate-to"):
+            primary_status = juju_utils.get_application_status( \
+                    status.get("subordinate-to")[0])
+            leader = None
+            for unit in primary_status["units"]:
+                if primary_status["units"][unit].get('leader'):
+                    leader = unit
+        leader_unit = zaza.model.get_unit_from_name(leader)
+        assert leader_unit.workload_status == "active"
```
Comment (thedac) [inline zaza/openstack/charm_tests/hacluster/tests.py]: We don't need this check. And in fact, you can remove the leader_unit as we never use it for anything else.

File: zaza/openstack/charm_tests/hacluster/tests.py [hunk 4]
```diff
@@ -0,0 +1,104 @@
...
+        # Use the application it is subordinate-to to check workload status
+        if status.get("units") is None and status.get("subordinate-to"):
+            primary_status = juju_utils.get_application_status(
+                    status.get("subordinate-to")[0])
+            leader = None
+            for unit in primary_status["units"]:
+                if primary_status["units"][unit].get('leader'):
+                    leader = unit
+        leader_unit = zaza.model.get_unit_from_name(leader)
+        assert leader_unit.workload_status == "active"
```
Comment (thedac) [inline zaza/openstack/charm_tests/hacluster/tests.py]: Like above: We don't need this check. And in fact, you can remove the leader_unit as we never use it for anything else.

**Comment by xiywang** [comment]
@thedac Updated. Please help review again. Thanks.

File: zaza/openstack/charm_tests/hacluster/__init__.py [hunk 5]
```diff
@@ -0,0 +1,15 @@
+# Copyright 2018 Canonical Ltd.
```
Comment (thedac) [inline zaza/openstack/charm_tests/hacluster/__init__.py]: Update to 2019

File: zaza/openstack/charm_tests/hacluster/tests.py [hunk 6]
```diff
@@ -0,0 +1,121 @@
+#!/usr/bin/env python3
+
+# Copyright 2018 Canonical Ltd.
```
Comment (thedac) [inline zaza/openstack/charm_tests/hacluster/tests.py]: Update to 2019

File: zaza/openstack/charm_tests/hacluster/tests.py [hunk 7]
```diff
@@ -0,0 +1,121 @@
...
+                zaza.model.run_action(subordinate, "resume")
+                zaza.model.block_until_unit_wl_status(leader, "active")
+
+        _states = {"hacluster": {
+            "workload-status": "active",
+            "workload-status-message": "Unit is ready and clustered"}}
+        zaza.model.wait_for_application_states(states=_states)
+        logging.debug('OK')
+
+    def _toggle_maintenance_and_wait(self, expected):
```
Comment (thedac) [inline zaza/openstack/charm_tests/hacluster/tests.py]: Let's add docstrings to this method.

**Comment by xiywang** [comment]
@thedac Done. Thanks for review.

**Pull request merged**